keps/sig-auth/2799-token-controller-deprecation/README.md

```diff
@@ -0,0 +1,474 @@
+- [Implementation History](#implementation-history)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+- [Infrastructure Needed (Optional)](#infrastructure-needed-optional)
+<!-- /toc -->
+
+## Release Signoff Checklist
+
+Items marked with (R) are required _prior to targeting to a milestone /
+release_.
+
+- [x] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [x] (R) KEP approvers have approved the KEP status as `implementable`
+- [x] (R) Design details are appropriately documented
+- [x] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input (including test refactors)
+- [x] (R) Graduation criteria is in place
+- [x] (R) Production readiness review completed
+- [x] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+This KEP proposes a namespace-scoped and service-account-scoped binary label to
+allow users to opt in/out the provision of secret-based service account tokens
+in Token Controller. In addition, it sketches out the phases to deprecate Token
+Controller.
+
+## Motivation
+
+As BoundServiceAccountTokenVolume is GA in 1.22, pods’ service account tokens
+would be obtained via TokenRequest API and stored as projected volume. This
+change obviates the need for Token Controller’s reconciling of secret-based
+service account tokens. The secret-based tokens are [not secure by design](https://github.com/kubernetes/enhancements/tree/master/keps/sig-auth/1205-bound-service-account-tokens#background)
+and the token controller is [fragile by design in some cases](https://github.com/kubernetes/kubernetes/issues/98474)
+where it is unable to handle the churns between secrets and service account
+controller loops.
+
+### Goals
+
+- Provides a knob to opt in/out the generation of secret-based token for a
+  service account
+- Provides a knob to opt in/out the generation of secret-based token for all
+  service accounts in a namespace
+- Deprecate Token Controller
+- No Forever Token in Kubernetes
+
+### Non-Goals
+
+## Proposal
+
+With the goal to deprecate Token Controller,
+
+- Phase I: At release N, `tokencontroller.kubernetes.io/reconcile` is `true` by default
+  to align with existing behavior of Token Controller. To disable the
+  generation of secret-based tokens, users are required to specify
+  `tokencontroller.kubernetes.io/reconcile=false` for the service accounts or
```

Comment:
AFAICT: we usually don't mention the term “controller” in annotation names etc. I'd expect to see this named something like:
`serviceaccount.kubernetes.io/generate-secret` or `serviceaccount.kubernetes.io/reconcile-secret`.

In terms of people consuming the API, the existence of controllers is an implementation detail they don't need to know about. AIUI the feature that's changing is Secret-based tokens for ServiceAccounts, whether that's done by the kube-controller-manager or magic elves.

(so, I'd name it accordingly).

Reply:
good suggestion. changed it to `serviceaccount.kubernetes.io/reconcile-legacy-token`